### ISSUE TITLE
Sub-command showing help message for sub-sub-command

### ISSUE THREAD

**Issue by psanford** [codegangsta/cli#163]
Consider the following program:

```
package main

import (
	"os"

	"github.com/codegangsta/cli"
)

func main() {
	app := cli.NewApp()
	app.Commands = []cli.Command{
		{
			Name: "a",
			Subcommands: []cli.Command{
				{
					Name: "a1",
				},
				{
					Name: "a2",
				},
			},
		},
		{
			Name: "b",
			Subcommands: []cli.Command{
				{
					Name: "b1",
				},
				{
					Name: "b2",
				},
			},
		},
	}
	app.Run(os.Args)
}
```

I would expect the output for `./bin/cliexample a` to show the help text of the `a` command. But instead it shows the help text of `a1`: 

```
 $ ./bin/cliexample a
NAME:
   a1 - 

USAGE:
   command a1 [arguments...]
```

However if you give it the help command it shows the correct help text:
```
 ./bin/cliexample a help
NAME:
   cliexample a - 

USAGE:
   cliexample a [global options] command [command options] [arguments...]

VERSION:
   0.0.0

AUTHOR:
  Author - <unknown@email>

COMMANDS:
   a1
   a2
   help, h      Shows a list of commands or help for one command
   
GLOBAL OPTIONS:
   --help, -h   show help
```

**Comment by frellus** [comment]
I get the same behavior. I'm just starting to use this code, but will try to figure out a solution if I can and send a fix up

**Comment by jszwedko** [comment]
This seems to be working correctly in master, can you verify @psanford ?

**Comment by psanford** [comment]
I still experience the problem on master as of bf4a526f, what commit do you think fixed it?

**Comment by nrdufour** [comment]
I have the same issue. I've been trying to figure out a good way to solve it for a while.
It seems that the template `SubcommandHelpTemplate` is never used anywhere.
And I think that the implementation of `ShowCommandHelp` is rather strange or wrong.

**Comment by oleksandr** [comment]
Confirming the same issue form the recent master.

**Comment by pandrew** [comment]
Ran into this right know.

**Comment by jszwedko** [comment]
Apologies for the delay!

I'm not sure why I thought I was seeing the correct behaviour, but I was able to reproduce this on master. I'm still trying to wrap my head around the app/command/subcommand differences, but #172 appears to fix this issue. I'd like to refactor some of the duplication between `App` and `Command` away.

**Comment by ehazlett** [comment]
So oddly #172 appears to fix it if using `command help` or `command h` (it lists all subcommands) but shows help for the first subcommand only if using `command --help` or `command -h`.  Am I using it wrong?

```
ehazlett subcommands> ./subcommands 
NAME:
   ./subcommands - A new cli application

USAGE:
   ./subcommands [global options] command [command options] [arguments...]

VERSION:
   0.0.0

AUTHOR:
  Author - <unknown@email>

COMMANDS:
   parent       parent
   another      Another
   help, h      Shows a list of commands or help for one command
   
GLOBAL OPTIONS:
   --help, -h           show help
   --version, -v        print the version
   
ehazlett subcommands> ./subcommands parent help
NAME:
   ./subcommands parent - parent

USAGE:
   ./subcommands parent [global options] command [command options] [arguments...]

VERSION:
   0.0.0

AUTHOR:
  Author - <unknown@email>

COMMANDS:
   child-foo    Child Foo
   child-bar    Child Bar
   help, h      Shows a list of commands or help for one command
   
GLOBAL OPTIONS:
   --help, -h   show help
   
ehazlett subcommands> ./subcommands parent -h
NAME:
   child-foo - Child Foo

USAGE:
   command child-foo [arguments...]
```

**Comment by ehazlett** [comment]
Another side effect is the app info appears to be reset as well (name, version, author, email).

**Comment by wricardo** [comment]
I've just experienced this same issue. I hope it gets fixed soon.

**Comment by DanielBryan** [comment]
Yep, same issue here.

This works:

    app subcommand help

But this will show the help of whichever sub-sub-command comes first in the list:

    app subcommand

**Comment by jszwedko** [comment]
Sorry for the delay -- appears to be fixed by #186.
